File: CR12/Lectures13-14/unfolding.m
function T = unfolding(X, k)

	nrows = size(X,k);
	ncols = numel(X) / nrows;
	T = zeros(nrows, ncols);
	
	dimsize = size(X);
	
	offset = dimsize;
	
	offset(k) = [];
	
	offset = [1 offset];
	
	for i=2:length(offset)
		offset(i) = offset(i) * offset(i-1);
	end
	
	for icol=1:ncols
		tindex = ones(1, ndims(X));
		id = icol-1;
		for i=ndims(X):-1:k+1
			tindex(i) += floor(id/offset(i-1));
			id = mod(id, offset(i-1));
		end
		for i=k-1:-1:1
			tindex(i) += floor(id/offset(i));
			id = mod(id, offset(i));
		end
		
			tmp2 = num2cell(tindex(k+1:ndims(X)));
			tmp1 = num2cell(tindex(1:k-1));
			T(:,icol)=X(tmp1{:},:,tmp2{:});
	%{	
		if k==1
			tmp = num2cell(tindex(k+1:ndims(X)));
			T(:,icol)=X(:,tmp{:});
		elseif k==ndims(X)
			tmp = num2cell(tindex(1:k-1));
			T(:,icol)=X(tmp{:},:);
		else
			T(:,icol)=X(tindex(1:k-1),:,tindex(k+1:ndims(X)));
		end
	%}
		
	end



end

File: CR12-2024/Lecture14/unfolding.py
import numpy as np
def unfolding(X, k):
    shape = X.shape
    ndims = len(shape)
    nrows = shape[k]
    ncols = X.size//nrows

    Xk = np.zeros((nrows, ncols))

    offset = np.asarray(shape)
    offset[k] = 1

    t=1
    for i in range(ndims):
        oldt = t
        t = t * offset[i]
        offset[i] = oldt
    
    for icol in range(ncols):
        current_id = icol
        index = np.zeros(ndims, dtype=int)
    
        for i in reversed(range(ndims)):
            if (i != k):
                index[i] = current_id // offset[i]
                current_id = current_id % offset[i]
    
        for i in range(nrows):
            index[k] = i
            Xk[i][icol] = X[tuple(index)]

    return Xk
    


X = np.zeros((3,3,3))
for i in range(3):
    for j in range(3):
        for k in range(3):
            X[i][j][k] = (i+1) + (j+1)**2 + (k+1)**3


print(unfolding(X,0))
print(unfolding(X,1))
print(unfolding(X,2))

